lib/libutee/arch/arm/user_ta_entry.c

```diff
@@ -33,42 +35,118 @@ extern struct ta_head ta_head;
 
 uint32_t ta_param_types;
 TEE_Param ta_params[TEE_NUM_PARAMS];
-struct __init_fini_info __init_fini_info;
 struct __elf_phdr_info __elf_phdr_info;
 
-void __utee_call_elf_init_fn(void)
+struct phdr_info {
+	struct dl_phdr_info info;
+	TAILQ_ENTRY(phdr_info) link;
+};
+
+static TAILQ_HEAD(phdr_info_head, phdr_info) _phdr_info_head =
+		TAILQ_HEAD_INITIALIZER(_phdr_info_head);
+/*
+ * Keep track of how many modules have been initialized so that subsequent
+ * dlopen() calls will not run the same initializers again
+ */
+static size_t _num_mod_init;
+
+static int _init_iterate_phdr_cb(struct dl_phdr_info *info,
+				 size_t size __unused, void *data)
+{
+	struct phdr_info *qe = NULL;
+	size_t *count = data;
+
+	qe = malloc(sizeof(*qe));
+	if (!qe) {
+		EMSG("init/fini: out of memory");
+		abort();
+	}
+	qe->info = *info;
+	TAILQ_INSERT_TAIL(&_phdr_info_head, qe, link);
+	(*count)++;
+	return 0;
+}
+
+static void _get_fn_array(struct dl_phdr_info *info, Elf_Sword tag_a,
+			  Elf_Sword tag_s, void (***fn)(void), size_t *num_fn)
 {
-	size_t n = __init_fini_info.size;
+	const Elf_Phdr *phdr = NULL;
+	Elf_Dyn *dyn = NULL;
+	size_t num_dyn = 0;
 	size_t i = 0;
 	size_t j = 0;
 
-	for (i = 1; i <= n; i++) {
-		/* Reverse order: dependencies first */
-		struct __init_fini *ifs = &__init_fini_info.ifs[n - i];
+	for (i = 0; i < info->dlpi_phnum; i++) {
+		phdr = info->dlpi_phdr + i;
+		if (phdr->p_type != PT_DYNAMIC)
+			continue;
+		num_dyn = phdr->p_memsz / sizeof(Elf_Dyn);
+		dyn = (Elf_Dyn *)(phdr->p_vaddr + info->dlpi_addr);
+		for (j = 0; j < num_dyn; j++) {
+			if (*fn && *num_fn)
+				break;
+			if (dyn->d_tag == DT_NULL) {
+				break;
+			} else if (dyn->d_tag == tag_a) {
+				*fn = (void (**)(void))(dyn->d_un.d_ptr +
+							info->dlpi_addr);
+			} else if (dyn->d_tag == tag_s) {
+				*num_fn = dyn->d_un.d_val / sizeof(Elf_Addr);
+			}
+			dyn++;
+		}
+	}
+}
 
-		if (!(ifs->flags & __IFS_INIT_HAS_RUN))
-			for (j = 0; j < ifs->init_size; j++)
-				ifs->init[j]();
+void __utee_call_elf_init_fn(void)
+{
+	void (**fn)(void) = NULL;
+	size_t num_mod = 0;
+	size_t num_fn = 0;
+	size_t mod = 0;
+	size_t i = 0;
+	struct phdr_info *qe = NULL;
+	struct phdr_info *qe2 = NULL;
+
+	dl_iterate_phdr(_init_iterate_phdr_cb, &num_mod);
 
-		ifs->flags |= __IFS_INIT_HAS_RUN;
+	/* Reverse order: dependencies first */
+	TAILQ_FOREACH_REVERSE(qe, &_phdr_info_head, phdr_info_head, link) {
+		if (mod == num_mod - _num_mod_init)
+			break;
+		_get_fn_array(&qe->info, DT_INIT_ARRAY, DT_INIT_ARRAYSZ, &fn,
+			      &num_fn);
+		for (i = 0; i < num_fn; i++)
+			fn[i]();
+		fn = NULL;
+		num_fn = 0;
+		mod++;
 	}
+	_num_mod_init += mod;
+
+	TAILQ_FOREACH_SAFE(qe, &_phdr_info_head, link, qe2)
+		free(qe);
```

Comment:
This works, but looks a bit scary with the dangling pointers. Normally I'd write it like:
``` C
TAILQ_REMOVE(&_phdr_info_head, qe, link);
free(qe);
```
and skipping the cleaning `TAILQ_INIT()` below.

Reply:
Agreed.